208 fractional epochs

Nacho114 opened the pull request:
Issue: #208
In this PR we implement rounds which can be a fraction of an epoch (especially important for large datasets)

This is still work in progress, it seems to work for roundDuration <= 1 (original case + fractional case)

(Note I also address  suggestions on async by @martinjaggi such as handler->buffer)

@martinjaggi @s314cy 

I am worried about the naming: in the buffer we also have `rounds` now. However, in that context, a round is incremented when the buffer is filled up, while related to the local training round, it is not exactly the same, and I think that the naming being identical but slightly different might cause confusion. Perhaps we should rename the rounds in the buffer as cycles or something else (e.g. currentBufferCycle, bufferCycle, cycle, ...)? 

TODO:

- [x] Debug the roundDuration > 1 case (even though it might not be used, it is good to understand the error).
- [x] Address naming issue.
- [x] Implement it for decentralized.
- [x] Address #210 as this overlaps a bit.

Nacho114 committed:
updated async for better clarity

Nacho114 committed:
added first version of fractional round, working for rounds <= 1, added tests, still need to debug the > 1 case

martinjaggi commented:
thanks, looking good!

it should be fine to still call it `round`, as it is the same unit basically. here it's how many batches a client will need to train locally, until it can sent it to server or peers (to then become part of the corresponding round of the buffer, if selected and not delayed)

Nacho114 commented:
yeah true, they are the same unit. But I think in the client, we should keep in mind that there are two "rounds", one, the current local training round, and then there is also the remote round which corresponds to the latest fetched model (i.e. in the remote server at what round was the buffer averaged). So we could distinguish them as round vs remoteRound. 

In the buffer and client we keep calling rounds, rounds, but in the client we have a separate name for the remote rounds.

martinjaggi commented:
in federated it's easiest to say the round number is dictated by the server, whenever a new averaging happens (buffer flush).

then when clients pull a model they also get the round metadata with the model, then do several local steps, and finally send back. when sending back it's crucial they attach the metadata of what time this update was originally based on (old model which was pulled). which the server (buffer) then uses to decide to drop it or not when received. 

let's use some terminology like `model update is based on round number T`, for these models going client->server

Nacho114 committed:
fixed fractional rounds bug

Nacho114 commented:
fractional rounds are now working , still need to implement it for decentralized before merging the PR.

Nacho114 commented:
@s314cy 
How are `postMetadata` and `getMetadataMap` used? I see them defined in fed client, but I can't see find where they are called?

Nacho114 committed:
added fractionalRounds to decentralised, made client abstract

s314cy commented:
> @s314cy 
> How are `postMetadata` and `getMetadataMap` used? I see them defined in fed client, but I can't see find where they are called? 

They are not being used yet. They were made for the future federated training informant to provide with stats shared by other nodes via the centralized server.

Nacho114 committed:
removed sync scheme code, rip

Nacho114 committed:
removed normal fit tf, and fixed fixDatset for titanic, was only working for cifar due to issue in data preproc

Nacho114 commented:
> weights buffer

you mean rename asyncWeightsBuffer rneamed to WeightsBuffer?

martinjaggi commented:
> > weights buffer
> 
> you mean rename asyncWeightsBuffer renamed to WeightsBuffer?

no it was only affecting the comment, not the code

Nacho114 commented:
> remove all epoch stuff, only use rounds?

We still use epochs for local training since there rounds dont matter. I could completely remove epochs and replace with round calls, but I would need to change a bit since I made this with only communication in mind.

martinjaggi commented:
ok will adapt local training later in a separate issue, not a high priority (it can also use rounds instead of epochs, config for users remains more clear then in both cases)

martinjaggi commented on mobile-browser-based-version/src/logic/communication/federated/federated_client.ts:
```diff
@@ -21,9 +21,16 @@ export class FederatedAsyncClient extends Client {
-    console.log('building Federated Async client')
     this.clientID = ''
-    this.round = -1 // The server starts at round 0, in the beginning we are behind
+    this.remoteModelRoundNumber = -1 // The server starts at round 0, in the beginning we are behind
```
rename `remoteModelRoundNumber` -> `modelUpdateIsBasedOnRoundNumber`

martinjaggi commented on mobile-browser-based-version/tests/logic/communication/client.test.ts:
```diff
@@ -0,0 +1,83 @@
+      expect(client._localRoundHasEnded(example.values[0], example.values[1], example.values[2], example.values[3], example.values[4])).equal(example.output)
+    })
+  })
+})
```
very nice work on the tests!

Nacho114 committed:
removed epoch callbacks and improved logic of code

Nacho114 commented:
Note that to keep track of rounds we need to keep track of epochs. 

After reading your comments I realize I can improve the training manager a bit further, I will do these changes tomorrow morning and finalize this PR.

Nacho114 committed:
refactored trainer into local and distributed classes, cleaned logic and creater roundTracker class

Nacho114 commented:
After reading Martins comments and a discussion with @s314cy I've refactored the training classes.

1. I switched the names trainer and training_manager, since the primer was actually being the manager.
2. I made the trainer (previously training_manager) into an abstract class, and I made a separate LocalTrainer and DistributedTrainer classes that extend the trainer class that each perform local/distributed training. It's nice to separate these two into different classes since the trainer (previously training_manager) was quite packed. 
3. I deleted all unnecessary onEpoch callbacks, however for updating the UI graph it is necessary to call the onEpochEnd callback since onBatchEnd gives us no validation accuracy.
4. Keeping track of fractional rounds is not completely trivial, so I made a separate class for it, I explain in comments why it is necessary to keep track of the latest epoch in order to keep track of the current round. 
5. The result of the refactoring is that we now have small independent modules that do specific tasks and will hence be easily testable once @s314cy finishes re-working the loadData pipeline.

martinjaggi commented on mobile-browser-based-version/src/logic/training/round_tracker.ts:
```diff
@@ -0,0 +1,68 @@
+   * Returns true if a local round has ended, false otherwise.
+   *
+   * @remark
+   * Batch is the current batch, this goes from 1, ... , batchSize*trainSize.
```
typo here, batch size doesn't matter, but you meant trainsize/batchsize. anyway would be nicer to just let batch count to infinity, not reset to zero every epoch

Nacho114 committed:
removed epochs from roundTracker and used cumulative batches instead

Nacho114 committed:
removed redundant class + added types on files that were worked on

Nacho114 committed:
turned platformed into enum and cleaned to reflect changes

Nacho114 committed:
added better comments and cleaning

s314cy commented on mobile-browser-based-version/server/tests/logic/federated/async_weights_buffer.test.ts:
```diff
@@ -0,0 +1,69 @@
+describe('AsyncWeightHandler tests', () => {
+  it('add weight with old time stamp returns false', async () => {
+    const t0 = -1 //
+    const asyncWeightBuffer = new AsyncWeightsBuffer(taskId, bufferCapacity, mockAggregateAndStoreWeights)
```
`asyncWeightsBuffer`

s314cy commented on mobile-browser-based-version/server/tests/logic/federated/async_weights_buffer.test.ts:
```diff
@@ -0,0 +1,69 @@
+  mockUpdatedWeights = _weights
+}
+
+describe('AsyncWeightHandler tests', () => {
```
`'AsyncWeightsHandler tests'`

s314cy commented on mobile-browser-based-version/server/tests/router/federated.test.ts:
```diff
@@ -30,15 +30,15 @@ function disconnectHeader (platformID: string, taskID: string, clientID: string)
 
-function postAsyncWeightHeader (platformID: string, taskID: string, clientID: string) {
-  return `/${platformID}/async/weights/${taskID}/${clientID}`
+function postWeightHeader (platformID: string, taskID: string, clientID: string) {
```
`postWeightsHeader`

s314cy commented on mobile-browser-based-version/server/src/router/router.ts:
```diff
@@ -40,18 +40,9 @@ federatedRouter.get('/', (req, res) => res.send('FeAI server'))
 federatedRouter.get('/disconnect/:task/:id', handlers.disconnect)
 
-federatedRouter.get('/selection/:task/:id', handlers.selectionStatus)
+federatedRouter.post('/weights/:task/:id', handlers.postWeights)
```
:+1:

s314cy commented on mobile-browser-based-version/src/components/sidebar/Settings.vue:
```diff
@@ -191,7 +191,7 @@ export default defineComponent({
     ]),
     toggleIndexedDB () {
-      this.setIndexedDB(!this.$store.state.useIndexedDB && window.indexedDB)
+      this.setIndexedDB(!!(!this.$store.state.useIndexedDB && window.indexedDB))
```
the commit function in `store.js` already converts its argument to boolean, to avoid writing this trick in every call

s314cy commented on mobile-browser-based-version/src/logic/communication/client.ts:
```diff
@@ -1,57 +1,53 @@
-  serverURL: any;
-  task: any;
-  constructor (serverURL, task) {
+export abstract class Client {
```
:+1:

s314cy commented on mobile-browser-based-version/src/logic/communication/client.ts:
```diff
@@ -1,57 +1,53 @@
-  async connect (epochs?): Promise<any> {
-    throw new Error('Abstract method')
-  }
+  abstract connect (epochs?:number): Promise<any>
```
:+1:

s314cy commented on mobile-browser-based-version/src/logic/communication/client.ts:
```diff
@@ -1,57 +1,53 @@
-  async onTrainEndCommunication (model, trainingInformant) {
-    trainingInformant.addMessage('Training finished.')
-  }
+  abstract onTrainEndCommunication (model, trainingInformant: TrainingInformant): Promise<void>
```
I think the callback methods should not be abstract but rather be some sort of `default` methods, since the `Client` abstract class may include some routine common to every subclasses, such as displaying the RAM usage

Nacho114 replied:
The RAM usage should be displayed inside the onRoundEnd call backs in the trainer? If there are no common subroutines, is it not better to keep it abstract?

s314cy commented on mobile-browser-based-version/src/logic/communication/client.ts:
```diff
@@ -1,57 +1,53 @@
+   * Init holder for federated, we need to fetch latest server round
+   * and model before training.
+   */
+  async updateModelAndRoundFromServer () {}
```
I'm not sure why this is in `Client`, as this is typically not used by `DecentralizedClient`

Nacho114 replied:
you are right, it's also no longer used due to refactoring so I will delete it.

s314cy commented on mobile-browser-based-version/src/logic/communication/client_builder.ts:
```diff
@@ -1,25 +1,18 @@
+export function getClient (platform: Platform, task: Task, password: string): Client {
   switch (platform) {
-    case 'deai':
+    case Platform.decentralized:
```
:1st_place_medal:

s314cy commented on mobile-browser-based-version/src/logic/communication/decentralised/decentralised_client.ts:
```diff
@@ -158,21 +160,17 @@ export class DecentralisedClient extends Client {
    * Disconnection process when user quits the task.
    */
-  disconnect () {
+  async disconnect () {
```
we can discuss the `async`/`await` policy later on @Nacho114

Nacho114 replied:
yes, this is important, it's not clear to me what it is right now

s314cy commented on mobile-browser-based-version/src/logic/task_definition/base/task.ts:
```diff
@@ -1,6 +1,56 @@
 import * as tf from '@tensorflow/tfjs'
 import * as memory from '../../memory/model_io'
 
+class DataExample {
```
:+1:

s314cy commented on mobile-browser-based-version/src/logic/training/distributed_trainer.ts:
```diff
@@ -0,0 +1,96 @@
+/**
+ * Class whose role is to train a model in a distributed way with a given dataset.
+ */
+export class DistributedTrainer extends Trainer {
```
great! this simplifies my current work for #191

s314cy commented on mobile-browser-based-version/src/logic/training/round_tracker.ts:
```diff
@@ -0,0 +1,65 @@
+ * The role of the RoundTracker is to keep track of when a roundHasEnded using the current batch number. The batch in the RoundTracker is cumulative whereas
+ * in the onBatchEnd it is not (it resets to 0 after each epoch).
+ */
+export class RoundTracker {
```
:+1:

s314cy commented on mobile-browser-based-version/src/logic/training/trainer.ts:
```diff
@@ -1,147 +1,224 @@
+  useIndexedDB: boolean;
+  stopTrainingRequested: boolean = false;
+  model: tf.LayersModel;
+  data: any; // TODO: use class type @s314cy once you have created the dataset class
```
:+1:

Nacho114 committed:
added further refactoring, private/protected classes, added builders to clean and simplify code

Nacho114 committed:
adapated test to changes and fixed typos mentioned in comments of pr
